spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -93,7 +94,8 @@
 
 	private final RedisTemplate<String, RedisLock> redisTemplate;
 
-	private final ThreadLocal<List<RedisLock>> threadLocks = new ThreadLocal<List<RedisLock>>();
+	private final ThreadLocal<Set<RedisLock>> weakThreadLocks = new ThreadLocal<Set<RedisLock>>();
+	private final ThreadLocal<List<RedisLock>> hardThreadLocks = new ThreadLocal<List<RedisLock>>();
```

Comment:
Well, I think we should fix the `hashCode` just for the immutable properties. We can't lock the same `lock` from the second Thread (or even the separate JVM process) because the guaranty on the Redis.
You second concern does not make sense without extracting variables. Although we can document that we get deal with `WeakRererence`s.

To be honest I don't feel confident in this stuff. Need @garyrussell input.

Anyway, @kiakimov , thank you very much for the help and contribution!